ow_power_system/src/power_system_node.cpp

```diff
@@ -199,66 +221,97 @@ double PowerSystemNode::generateVoltageEstimate()
 
   // Voltage estimate based on pseudorandom noise and moving range
-  uniform_real_distribution<double> m_voltage_dist(min_V, max_V);
-  return m_voltage_dist(m_random_generator);
+  uniform_real_distribution<double> voltage_dist(min_V, max_V);
+  return voltage_dist(m_random_generator);
 }
 
-void PowerSystemNode::injectFault(const string& power_fault_name, bool& fault_activated,
-                                  const vector<map<MessageId, Datum<double>>>& sequence, size_t& sequence_index,
-                                  double& power, double& voltage, double& temperature)
+void PowerSystemNode::injectFault (const string& fault_name,
+                                   bool& fault_activated,
+                                   const PrognoserVector& sequence,
+                                   size_t& index,
+                                   double& wattage,
+                                   double& voltage,
+                                   double& temperature)
 {
   bool fault_enabled = false;
-  bool success = ros::param::getCached("/faults/" + power_fault_name, fault_enabled);
-  if (!success)
+
+  // Do nothing unless the specified fault has been injected.
+  if (! ros::param::getCached("/faults/" + fault_name, fault_enabled)) {
     return;
+  }
 
   if (!fault_activated && fault_enabled)
   {
-    ROS_INFO_STREAM(power_fault_name << " activated!");
-    sequence_index = 0;
+    ROS_INFO_STREAM(fault_name << " activated!");
+    index = 0;
     fault_activated = true;
   }
   else if (fault_activated && !fault_enabled)
   {
-    ROS_INFO_STREAM(power_fault_name << " de-activated!");
+    ROS_INFO_STREAM(fault_name << " de-activated!");
     fault_activated = false;
   }
 
   if (fault_activated && fault_enabled)
   {
-    auto data = sequence[sequence_index++ % sequence.size()];  // TODO: for now we replay the sequence once it reaches
-                                                               // the This is unlikely the end case. Re-visit
-
-    auto m = 1.0 / m_power_node_processing_rate;  // multiplier that should be applied since this profile data are given in seconds
-                                                  // but the processing rate can be lesser or higher than that
-    power += m * data[MessageId::Watts];
-    voltage += m * data[MessageId::Volts];
-    temperature += m * data[MessageId::Centigrade];
+    // TODO: Unspecified how to handle end of fault profile, which is
+    // unlikely.  For now, reuse the last entry.
+    if (index + PROFILE_INCREMENT >= sequence.size()) {
+      if (! m_fault_profile_exhausted[fault_name]) {
+        ROS_WARN_STREAM(fault_name
+                        << ": reached end of fault profile, reusing last entry.");
+        m_fault_profile_exhausted[fault_name] = true;
+      }
+      index = sequence.size() - 1;
+    }
+    else index += PROFILE_INCREMENT;
+
+    auto data = sequence[index];
+    wattage = std::min (wattage + data[MessageId::Watts], MAX_GSAP_INPUT_WATTS);
```

Comment:
I think whatever causes the prognoser to behave erratically when power goes over 30W should be investigated in another ticket, if we don't have one for it already (Jira won't load for me right now, so I can't check).

I am also wondering if it would make sense to log a warning whenever `MAX_GSAP_INPUT_WATTS` is used instead of the actual mechanical power. Thoughts?

Reply:
Agreed on the first comment, but let me just ask Chetan first.

Excellent suggestion in the second comment: implemented, and I tested it by spoofing a high power value in a fault profile.